controllers/common/util.go

```diff
@@ -159,6 +176,57 @@ func Reverse(original []string) []string {
 	return reversed
 }
 
-func Namespacelize(resource string) string {
-	return strings.ReplaceAll(resource, "placeholder", constant.MasterNamespace)
+// Namespacelize adds the namespace specified
+func Namespacelize(resource, ns string) string {
+	return strings.ReplaceAll(resource, "placeholder", ns)
+}
+
+// Get ConfigMap of Common Services Maps
+func getCmOfMapCs(r client.Reader) *corev1.ConfigMap {
+	cmName := constant.CsMapConfigMap
+	cmNs := "kube-public"
+	csConfigmap := &corev1.ConfigMap{}
+	err := r.Get(context.TODO(), types.NamespacedName{Name: cmName, Namespace: cmNs}, csConfigmap)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("common service configmap kube-public/common-service-maps is not found: %v", err)
+			return nil
+		}
+		klog.Errorf("failed to fetch configmap kube-public/common-service-maps: %v", err)
+		return nil
+	}
+	return csConfigmap
+}
+
+// GetMasterNs gets MasterNamespaces of deploying Common Services
+func GetMasterNs(r client.Reader) string {
+	operatorNs, err := GetOperatorNamespace()
+	if err != nil {
+		klog.Errorf("Getting operator namespace failed: %v", err)
+		return ""
+	}
+
+	csConfigmap := getCmOfMapCs(r)
+	commonServiceMaps := csConfigmap.Data["common-service-maps.yaml"]
+	var cmData csMaps
+	if err := utilyaml.Unmarshal([]byte(commonServiceMaps), &cmData); err != nil {
+		klog.Errorf("Failed to fetch data of configmap common-service-maps: %v", err)
+		return ""
+	}
+
+	var masterNs string
+	if cmData.DefaultCsNs == "" {
```

Comment:
Do we need to check the `DefaultCsNs`?

I believe if the operator Namespace can't be found from the configmap, the master namespace should be `ibm-common-services`.

`DefaultCsNs` is only used when the webhook replacing the registry namespace in the operandrequest.

Reply:
Please correct me if I am wrong for the following scenario:

Use the example above,
```
data:
  common-service-maps.yaml: |
    namespaceMapping:
      - requested-from-namespace: cp4iforFord
        map-common-service-namespace: common-services-ford
      - requested-from-namespace: cp4iforToyota
        map-common-service-namespace: common-services-toyota
    defaultCsNs: ibm-common-services
```

The customer would like to specify Common Services installed in `common-services-ford` and `common-services-toyota` for `cp4iforFord` and `cp4iforToyota` respectively.

And there also exists other `cp4iforHonda` and `cp4iforBMW` which customer would like to use the Common Services installed in `DefaultCsNs` called `common-services-cars` for.

In the above case, I think we need to install Common Services in `common-services-cars` and once the operator Namespace can't be found from the configmap, webhook will still replacing registry namespace in the operandrequest by `DefaultCsNs`

Reply:
In your case, I believe the data in the configmap should be 
```
data:
  common-service-maps.yaml: |
    namespaceMapping:
      - requested-from-namespace: cp4iforFord
        map-common-service-namespace: common-services-ford
      - requested-from-namespace: cp4iforToyota
        map-common-service-namespace: common-services-toyota
      - requested-from-namespace: cp4iforHonda
        map-common-service-namespace: common-services-cars
      - requested-from-namespace: cp4iforBMW
        map-common-service-namespace: common-services-cars
    defaultCsNs: ibm-common-services
```

From my understanding,  `defaultCsNs` is the registry namespace name that should be replaced by the webhook.

Considering two cases:
1. How about users want to deploy a group of common services in common-services-ford for cp4iforFord and deploy another group of common services in the `ibm-common-services`?
2. How to handle the operand requests generated from CP4MCM, if the webhook replaces all the registry namespaces?

Reply:
Just Pin me on slack anytime to discuss this.

Original--------------------------------------------------------
Sorry, I made some mistakes on configmap in the example.

And I understand that we hope to minimize the perception for our adopters which means the registry namespace in operand request will usually be `ibm-common-services`. Please correct me if I am wrong here.

I agree that it makes sense that the configmap would point out the requestedNs as `common-services-cars` for both Honda and BMW.
and if there is a new service installed for example CP4MCM, and user did not update the configmap, the group of common services will be deployed into `ibm-common-services`. But what if they want all other services use Common services in `common-services-cars`?

/*
Therefore, there is a tradeoff that if user want to set defaultNs to `common-services-cars`, it will conflict with minimizing perception for adopters which means operand requests created by them will have registry namespace `common-services-cars`.
*/

From my perspective, if we are able to let adopters make changes on registry namespace. I wonder that whether there is a way to simplify the configmap like this
```
data:
  common-service-maps.yaml: |
    namespaceMapping:
      - requested-from-namespace: cp4iforFord
        map-common-service-namespace: common-services-ford
      - requested-from-namespace: cp4iforToyota
        map-common-service-namespace: common-services-toyota
    defaultCsNs: common-services-cars
```

For the two cases you mentioned,

ConfigMap for first case: `common-services-ford` for `cp4iforFord`, and others will remain in `ibm-common-services`(it could be changed to `common-services-cars`)
```
data:
  common-service-maps.yaml: |
    namespaceMapping:
      - requested-from-namespace: cp4iforFord
        map-common-service-namespace: common-services-ford
    defaultCsNs: ibm-common-services #common-services-cars
```

ConfigMap for second case: All the operand request without `opreq.Namespace` `cp4iforFord` and `cp4mcmforToyota` will remain `registryNs` at `ibm-common-services` or change `registryNs` to `common-services-cars`.  
```
data:
  common-service-maps.yaml: |
    namespaceMapping:
      - requested-from-namespace: cp4iforFord
        map-common-service-namespace: common-services-ford
      - requested-from-namespace: cp4mcmforToyota
        map-common-service-namespace: common-services-Toyota
    defaultCsNs: ibm-common-services #common-services-cars
```